{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# JBrownies\n",
    "### Script to parse .gff file and retrieve metrics\n",
    "James, Nathalia, Michael\n",
    "10.30.2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3                                                                                                    \n",
    "\n",
    "#############                                                                                                             \n",
    "# JBrownies                                                                                                               \n",
    "# script name: GFF_PARSER_FINAL.py.py                                                                           \n",
    "# goal: parse gff3 file, retrieve information                                                                     \n",
    "# input files and extensions: Aspergillus nidulans .gff3 file                                                             \n",
    "# status: finished!                                                                                                       \n",
    "#############                                                                                                             \n",
    "\n",
    "import sys\n",
    "import re\n",
    "\n",
    "### Part1 - Data structure                                                                                                \n",
    "## strategy: dict - mrna - gff_type - feature id                                                                          \n",
    "# this will allow us to retrieve detailed info per gene                                                                   \n",
    "\n",
    "gff_file = open(sys.argv[1], 'r')\n",
    "\n",
    "dict_mrna_exon_ids = {}\n",
    "\n",
    "for line in gff_file:\n",
    "        if line.startswith('#'):\n",
    "                continue\n",
    "        else:\n",
    "                columns = line.rstrip().split('\\t')\n",
    "                #print(columns)                                                                                           \n",
    "                if columns[2] == 'exon' or columns[2] == 'five_prime_UTR' or columns[2] == 'three_prime_UTR':\n",
    "                        match = re.search(r\"ID=(\\w+);.*?Parent=(\\w+)\", columns[8])\n",
    "                        featureid = match.group(1)\n",
    "                        mrnaid = match.group(2)\n",
    "                        gff_type = columns[2]\n",
    "                        #print(exonid)                                                                                    \n",
    "                        #print(mrnaid)                                                                                    \n",
    "                        if not mrnaid in dict_mrna_exon_ids:\n",
    "                                dict_mrna_exon_ids[mrnaid]={} # creating dict for mrna                                    \n",
    "                                if not gff_type in dict_mrna_exon_ids[mrnaid]: # asking if our targets in mrna {}         \n",
    "                                        dict_mrna_exon_ids[mrnaid][gff_type]=[]      # if not seen it, create [] to hold \\\n",
    "them                                                                                                                      \n",
    "                                        dict_mrna_exon_ids[mrnaid][gff_type].append(featureid) # put target on []         \n",
    "                                else:\n",
    "                                        dict_mrna_exon_ids[mrnaid][gff_type].append(featureid) # if seen targets before i\\\n",
    "n {}, add to []                                                                                                           \n",
    "                        elif gff_type in dict_mrna_exon_ids[mrnaid]: # if mrna in {} and gfftype in {}                    \n",
    "                                dict_mrna_exon_ids[mrnaid][gff_type].append(featureid) # add the featureid (= matches) to\\\n",
    " []                                                                                                                       \n",
    "                        else:\n",
    "                                dict_mrna_exon_ids[mrnaid][gff_type]=[]\n",
    "                                dict_mrna_exon_ids[mrnaid][gff_type].append(featureid)\n",
    "\n",
    "### Part2 - Report of features per gene                                                                                   \n",
    "\n",
    "total_numb_exon = 0\n",
    "total_numb_genes = 0\n",
    "numb_genes_5utr = 0\n",
    "numb_genes_3utr =0\n",
    "\n",
    "#print(\"Features per gene\")                                                                                               \n",
    "for k in dict_mrna_exon_ids:\n",
    "        #print(k)                                                                                                         \n",
    "        total_numb_exon += len(dict_mrna_exon_ids[k]['exon'])\n",
    "        total_numb_genes += 1\n",
    "        for key2 in dict_mrna_exon_ids[k]:\n",
    "                #print('  ',key2,'\\t',len(dict_mrna_exon_ids[k][key2]))                                                   \n",
    "                #print(key2)                                                                                              \n",
    "                if key2 == 'five_prime_UTR':\n",
    "                        numb_genes_5utr +=1\n",
    "                if key2 == 'three_prime_UTR':\n",
    "                        numb_genes_3utr +=1\n",
    "                        \n",
    "### Part3 - Length of exons                                                                                               \n",
    "\n",
    "gff_file = open(sys.argv[1], 'r')\n",
    "\n",
    "length_exons = 0\n",
    "\n",
    "for line in gff_file:\n",
    "        if line.startswith('#'):\n",
    "                continue\n",
    "        else:\n",
    "                columns = line.rstrip().split('\\t')\n",
    "                #print(columns)                                                                                           \n",
    "                if columns[2] == 'exon':\n",
    "                        begin_frame = int(columns[3])\n",
    "                        end_frame = int(columns[4])\n",
    "                        length_exons += ((end_frame+1) - begin_frame)\n",
    "#print(length_exons)                                                                                                      \n",
    "#print(length_exons/total_numb_genes)\n",
    "\n",
    "\n",
    "### Part4 - Totals                                                                                                        \n",
    "\n",
    "gff_file = open(sys.argv[1], 'r')\n",
    "\n",
    "gene_dict = {}\n",
    "other_dict = {}\n",
    "\n",
    "for line in gff_file:\n",
    "        if line.startswith('#'):\n",
    "                continue\n",
    "        else:\n",
    "                columns = line.rstrip().split('\\t')\n",
    "                if not columns[2] in other_dict:\n",
    "                        other_dict[columns[2]] =1\n",
    "                else:\n",
    "                        other_dict[columns[2]] +=1\n",
    "                if columns[2] == 'gene':\n",
    "                    if not \"total_genes\" in gene_dict:\n",
    "                        gene_dict[\"total_genes\"] =1\n",
    "                    else:\n",
    "                        gene_dict[\"total_genes\"] +=1\n",
    "\n",
    "                elif columns[2] == 'exon':\n",
    "                    if not \"total_exons\" in gene_dict:\n",
    "                        gene_dict[\"total_exons\"] =1\n",
    "                    else:\n",
    "                        gene_dict[\"total_exons\"] +=1\n",
    "\n",
    "                elif columns[2] == 'mRNA':\n",
    "                    if not \"total_mRNA\" in gene_dict:\n",
    "                        gene_dict[\"total_mRNA\"] =1\n",
    "                    else:\n",
    "                        gene_dict[\"total_mRNA\"] +=1\n",
    "\n",
    "                                                                        \n",
    "### Part5 - Gene features per chromosome                                                                                  \n",
    "\n",
    "gff_file = open(sys.argv[1], 'r')\n",
    "\n",
    "chromo_dict = {}\n",
    "\n",
    "for line in gff_file:\n",
    "        if line.startswith('#'):\n",
    "                continue\n",
    "        else:\n",
    "                columns = line.rstrip().split('\\t')\n",
    "                match = re.search(r\"Chr[A-Z]+\", columns[0])\n",
    "                #print(match.group(0))                                                                                    \n",
    "                chrm_name = match.group(0)\n",
    "                #print(chrm_name)                                                                                         \n",
    "                if not chrm_name in chromo_dict:\n",
    "                        gcounter =1\n",
    "                        mcounter =0\n",
    "                        exoncounter =0\n",
    "                        threeprimeutr =0\n",
    "                        fiveprimeutr =0\n",
    "                        my_counts = [gcounter,mcounter,exoncounter,threeprimeutr,fiveprimeutr]\n",
    "                        chromo_dict[chrm_name] = my_counts\n",
    "                else:\n",
    "                        #print('else')                                                                                    \n",
    "                        if columns[2] == 'gene':\n",
    "                                gcounter += 1\n",
    "                        if columns[2] == 'mRNA':\n",
    "                                mcounter += 1\n",
    "                        if columns[2] == 'exon':\n",
    "                                exoncounter += 1\n",
    "                        if columns[2] == 'three_prime_UTR':\n",
    "                                threeprimeutr +=1\n",
    "                        if columns[2] == 'five_prime_UTR':\n",
    "                                fiveprimeutr +=1\n",
    "                        chromo_dict[chrm_name] = [gcounter,mcounter,exoncounter,threeprimeutr,fiveprimeutr]\n",
    "                        #print(chrm_name)                                                                                 \n",
    "\n",
    "\n",
    "### Part6 - Average gene length                                                                                           \n",
    "\n",
    "gff_file = open(sys.argv[1], 'r')\n",
    "\n",
    "length_genes = 0\n",
    "\n",
    "for line in gff_file:\n",
    "        if line.startswith('#'):\n",
    "                continue\n",
    "        else:\n",
    "                columns = line.rstrip().split('\\t')\n",
    "                #print(columns)                                                                                           \n",
    "                if columns[2] == 'gene':\n",
    "                        begin_frame = int(columns[3])\n",
    "                        end_frame = int(columns[4])\n",
    "                        length_genes += (end_frame+1)-begin_frame\n",
    "#print(length_exons)                                                                                                      \n",
    "#print(length_exons/total_numb_genes)                                                             \n",
    "\n",
    "\n",
    "######### PRINTING REPORT OF GFF FILE FEATURES                                                                            \n",
    "\n",
    "# genes:                                                                                                                  \n",
    "print(\"Total genes:\",gene_dict['total_genes'])\n",
    "\n",
    "# mRNA                                                                                                                    \n",
    "print(\"Total mRNA:\", gene_dict['total_mRNA'])\n",
    "\n",
    "# genes/chromosome:                                                                                                       \n",
    "print(\"Chromosome metrics:\")\n",
    "for k in chromo_dict:\n",
    "        print(\"  \",k)\n",
    "        print('     ',\"Total number of genes:\",chromo_dict[k][0])\n",
    "        print('     ',\"Total number of mRNA:\",chromo_dict[k][1])\n",
    "        print('     ',\"Total number of exons:\",chromo_dict[k][2])\n",
    "        print('     ',\"Total number of 3' UTR:\",chromo_dict[k][3])\n",
    "        print('     ',\"Total number of 4' UTR:\",chromo_dict[k][4])\n",
    "\n",
    "# avg gene_length:                                                                                                        \n",
    "gene_length_total = length_genes/total_numb_genes\n",
    "print(\"Average length of genes:\",\"{:.1f}\".format(gene_length_total),\" bp\")\n",
    "\n",
    "# exons/gene:                                                                                                             \n",
    "average_exons_gene = total_numb_exon/total_numb_genes\n",
    "print(\"Average number of exons per gene:\",\"{:.1f}\".format(average_exons_gene))\n",
    "\n",
    "# avg spliced transcript length:                                                                                          \n",
    "exon_length_total = length_exons/total_numb_genes\n",
    "print(\"Average length of spliced transcripts:\",\"{:.1f}\".format(exon_length_total),\" bp\")\n",
    "\n",
    "# avg exon_length:                                                                                                        \n",
    "avg_exon_length = length_exons/total_numb_exon\n",
    "print(\"Average length of exons:\",\"{:.1f}\".format(avg_exon_length),\" bp\")\n",
    "\n",
    "# avg intro_length                                                                                                        \n",
    "print(\"Average length of introns:\",\"{:.1f}\".format(length_introns/numb_introns),\" bp\")\n",
    "\n",
    "# % genes with 5' and 3' UTRs:                                                                                            \n",
    "#print(\"Number of genes with 5' UTR:\",numb_genes_5utr)                                                                    \n",
    "#print(\"Number of genes with 3' UTR:\",numb_genes_3utr)                                                                    \n",
    "print(\"Percent of genes with 5' UTR:\",\"{:.1%}\".format(numb_genes_5utr/total_numb_genes))\n",
    "print(\"Percent of genes with 3' UTR:\",\"{:.1%}\".format(numb_genes_5utr/total_numb_genes))\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make script executable, and run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ngraf@ip-172-30-0-209:/data/project$ ./GFF_PARSER_FINAL.py Asp_nidulans/Aspnid1.filtered_proteins.AspGD_genes.gff3 "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Total genes: 10680\n",
    "Total mRNA: 10680\n",
    "Chromosome metrics:\n",
    "   ChrIII\n",
    "      Total number of genes: 1261\n",
    "      Total number of mRNA: 1261\n",
    "      Total number of exons: 4117\n",
    "      Total number of 3' UTR: 416\n",
    "      Total number of 4' UTR: 448\n",
    "   ChrII\n",
    "      Total number of genes: 1461\n",
    "      Total number of mRNA: 1461\n",
    "      Total number of exons: 4631\n",
    "      Total number of 3' UTR: 418\n",
    "      Total number of 4' UTR: 444\n",
    "   ChrIV\n",
    "      Total number of genes: 1006\n",
    "      Total number of mRNA: 1006\n",
    "      Total number of exons: 3253\n",
    "      Total number of 3' UTR: 270\n",
    "      Total number of 4' UTR: 280\n",
    "   ChrI\n",
    "      Total number of genes: 1344\n",
    "      Total number of mRNA: 1344\n",
    "      Total number of exons: 4534\n",
    "      Total number of 3' UTR: 406\n",
    "      Total number of 4' UTR: 458\n",
    "   ChrVIII\n",
    "      Total number of genes: 1741\n",
    "      Total number of mRNA: 1741\n",
    "      Total number of exons: 5603\n",
    "      Total number of 3' UTR: 556\n",
    "      Total number of 4' UTR: 594\n",
    "   ChrVII\n",
    "      Total number of genes: 1593\n",
    "      Total number of mRNA: 1593\n",
    "      Total number of exons: 5115\n",
    "      Total number of 3' UTR: 470\n",
    "      Total number of 4' UTR: 507\n",
    "   ChrVI\n",
    "      Total number of genes: 1165\n",
    "      Total number of mRNA: 1165\n",
    "      Total number of exons: 3808\n",
    "      Total number of 3' UTR: 305\n",
    "      Total number of 4' UTR: 345\n",
    "   ChrV\n",
    "      Total number of genes: 1109\n",
    "      Total number of mRNA: 1109\n",
    "      Total number of exons: 3682\n",
    "      Total number of 3' UTR: 296\n",
    "      Total number of 4' UTR: 301\n",
    "Average length of genes: 1732.7  bp\n",
    "Average number of exons per gene: 3.3\n",
    "Average length of spliced transcripts: 1551.0  bp\n",
    "Average length of exons: 476.8  bp\n",
    "Average length of introns: 275.5  bp\n",
    "Percent of genes with 5' UTR: 29.1%\n",
    "Percent of genes with 3' UTR: 29.1%"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Another option is to direct the stdout using '>'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ngraf@ip-172-30-0-209:/data/project$ ./GFF_PARSER_FINAL.py Asp_nidulans/Aspnid1.filtered_proteins.AspGD_genes.gff3 > Aspni_gff_metrics.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
